arclightning_backend/src/main.rs

```diff
@@ -2,56 +2,75 @@ extern crate futures;
 extern crate hyper;
 #[macro_use]
 extern crate serde_derive;
-#[macro_use]
 extern crate serde_json;
 extern crate toml;
 
 use futures::future;
-use hyper::rt::{Future, Stream};
+use hyper::rt::Future;
 use hyper::service::service_fn;
 use hyper::{Body, Method, Request, Response, Server, StatusCode};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{self, ErrorKind, Read};
 use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
 
 // after this concept is further understood, will switch to 'Either'
 type ResponseFuture = Box<Future<Item = Response<Body>, Error = hyper::Error> + Send>;
 
-#[derive(Debug, Serialize, Deserialize, PartialEq)]
+// using PartialEq for unit tests
+// Using clone in a unit test atm.  Might not be necessary
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
 struct Game {
     name: String,
     description: String,
-    genre: Vec<String>,
+    genres: Vec<String>,
     thumbnail_path: PathBuf,
     exe_path: PathBuf,
 }
 
-fn router(request: Request<Body>) -> ResponseFuture {
-    let mut games_list: Vec<Game> = Vec::new();
-
+fn router(games_arc: &Arc<Mutex<HashMap<String, Game>>>, request: Request<Body>) -> ResponseFuture {
     let mut response = Response::new(Body::empty());
 
     match (request.method(), request.uri().path()) {
-        // TODO: get will send over games list
         (&Method::GET, "/api/v1/list_games") => {
-            // TODO: convert a vector of struct "Game" into a single json:
-            *response.body_mut() = Body::from("games will go here");
-        }
-        // TODO: post will probably figure out which game to launch?
-        (&Method::POST, "/echo") => {
-            *response.body_mut() = request.into_body();
+            *response.body_mut() =
+                Body::from(serde_json::to_string(&*games_arc.lock().unwrap()).unwrap());
         }
         _ => {
             *response.status_mut() = StatusCode::NOT_FOUND;
         }
     }
-
     Box::new(future::ok(response))
 }
 
+fn toml_to_hashmap(toml_filepath: PathBuf) -> Result<HashMap<String, Game>, io::Error> {
+    let mut games_toml = String::new();
+    let mut file = File::open(&toml_filepath)?;
+    file.read_to_string(&mut games_toml)?;
+
+    // error casting for homogeneous errors
+    toml::from_str(&games_toml).map_err(|e| io::Error::new(ErrorKind::Other, e.to_string()))
```

Comment:
Is`to_string` necessary here? I believe `io::Error::new` will accept any error type as input